hybrid/flow.py

```diff
@@ -623,34 +718,128 @@ def __repr__(self):
+        """Implement "converge on non-changing output" behavior:
+
+          - loop `max_iter` times, but bail-out earlier if output doesn't change
+            (over input) for `convergence` number of iterations
+
+          - each iteration starts with the previous result state
+
+        Input: relevant counters and I/O states.
+        Output: next input state and next counter values
+        """
+
+        input_energy = self.key(input_state)
+        output_energy = self.key(output_state)
+
+        logger.info("{name} Iteration(iterno={iterno}, output_state_energy={key})".format(
+            name=self.name, iterno=iterno, key=output_energy))
+
+        if output_energy == input_energy:
+            cnt -= 1
+        else:
+            cnt = self.convergence
+
+        return iterno + 1, cnt, output_state
+
     def next(self, state):
-        """Execute one blocking iteration of an instantiated :class:`Loop`."""
-        last = state
-        last_quality = self.key(last)
+        iterno = 0
         cnt = self.convergence
+        input_state = state
 
-        for iterno in range(self.max_iter):
-            state = self.runnable.run(state, executor=immediate_executor).result()
-            state_quality = self.key(state)
+        while True:
+            output_state = self.runnable.run(input_state, executor=immediate_executor).result()
 
-            logger.info("{name} Iteration(iterno={iterno}, best_state_quality={key})".format(
-                name=self.name, iterno=iterno, key=state_quality))
+            iterno, cnt, input_state = self.iteration_update(iterno, cnt, input_state, output_state)
 
-            if state_quality == last_quality:
-                cnt -= 1
-            else:
-                cnt = self.convergence
+            if self.max_iter is not None and iterno >= self.max_iter:
+                break
             if cnt <= 0:
                 break
 
-            last_quality = state_quality
+        return output_state
 
-        return state
+
+class LoopN(LoopUntilNoImprovement):
+    """Iterate `runnable` exactly `n` times, using each output as input in the
+    next iteration.
+    """
+
+    def __init__(self, runnable, n):
+        super(LoopN, self).__init__(
+            runnable=runnable, max_iter=n, convergence=n, key=lambda _: 0)
 
 
+class Loop(LoopUntilNoImprovement):
+    pass
+
 SimpleIterator = Loop
 
 
+class LoopWhileNoImprovement(LoopUntilNoImprovement):
+    """Iterates `runnable` until a state quality metric, defined by the `key`
+    function, shows no improvement for at least `max_tries` number of
+    iterations.
+
+    Args:
+        runnable (:class:`Runnable`):
+            A runnable that's looped over.
+
+        max_tries (int, optional, default=10):
+            Maximum number of times the `runnable` is run for the **same** input
+            state. On each improvement, the better state is used for the next
+            input state, and tries counter is reset.
```

Comment:
state, and `max_tries` counter is reset?

Reply:
No, "try counter", or "trial counter". A counter that counts tries, up to `max_tries`.